src/lib/queries/reward/getMyFindingsRewards.ts

```diff
@@ -19,34 +21,33 @@ const getMyFindingsRewardsQueryOptions = (
   queryFn: withApiErrorHandler(() => getMyFindingsRewards(params)),
 })
 
-export const useGetMyFindingsRewards = (params: GetMyFindingsRewardsParams) => {
-  const userId = useUserId()
-
-  return useQuery(getMyFindingsRewardsQueryOptions(userId, params))
-}
-
 export const prefetchGetMyFindingsRewards = async (
   userId: string,
   params: GetMyFindingsRewardsParams,
 ) => {
   const queryClient = getServerQueryClient()
-  await queryClient.prefetchQuery(
-    getMyFindingsRewardsQueryOptions(userId, params),
-  )
+  await Promise.all([
+    queryClient.prefetchQuery(getMyFindingsRewardsQueryOptions(userId, params)),
+    queryClient.prefetchQuery(getMyFindingsRewardsCountQueryOptions(userId)),
+  ])
 }
 
 const getMyFindingsRewardsCountQueryOptions = (userId: string | undefined) => ({
   queryKey: queryKeys.rewards.totalSize(userId).queryKey,
   queryFn: withApiErrorHandler(() => getMyFindingsRewardsCount()),
 })
 
-export const useGetMyFindingsRewardsCount = () => {
+export const useGetMyFindingsRewards = (
+  params: GetMyFindingsRewardsParams,
+): PaginatedResponse<MyFindingsReward[]> => {
   const userId = useUserId()
 
-  return useQuery(getMyFindingsRewardsCountQueryOptions(userId))
-}
+  const {data: countData} = useQuery(
+    getMyFindingsRewardsCountQueryOptions(userId),
+  )
 
-export const prefetchGetMyFindingsRewardsCount = async (userId: string) => {
-  const queryClient = getServerQueryClient()
-  await queryClient.prefetchQuery(getMyFindingsRewardsCountQueryOptions(userId))
+  return {
+    data: useQuery(getMyFindingsRewardsQueryOptions(userId, params)),
+    pageParams: {totalCount: countData?.count},
+  }
 }
```

Comment:
This is not ideal as you'll still have to write two useQuery hooks, that needs to be prefetched, invalidated etc... and also you're dropping the totalCount query data (isLoading, error, refetch...). What I meant is that we would merge those "count" hooks with "data" hooks and make the server actions to return total counts in addition to paginated response (basically the merging done in this file would happen on server).

Reply:
I thought about implementing that initially, but we lose the caching for the `totalCount` which in our use case will only be needed once to get the number of pages (and only refetched when query key gets invalidated). If the data is merged into one server action, for every param change (offset and sort), the `totalCount` db query will be called which may not be ideal. As for the query data, this can be added in however the current use case doesn't really require the `isLoading`, `error` and `refetch`.

As an alternative, I came across [this article](https://www.codemzy.com/blog/react-query-pagination-with-count) that encountered the same issue and still ended up using 2 queries but with only one data api/server action. We can consider this as an alternative to the current implementation, but this still raises the concern of having to fetch the `totalCount` every time. What do you think?

Reply:
I don't think it's bad to repeatedly query `totalCount`, that way we're sure that the `totalCount` is always consistent/up-to-date with the number of items that are actually returned and displayed on screen. I'd like this to be a general solution, that we could just easily use for all paginated queries without too much work. In that case, combining two queries (as it is now) seems like quite a lot of work that would need to be done repeatedly for every paginated query and also caching the total count and refetching just the actual data is somewhat fragile - imagine a user A queries the first page of some list, receives 10 items out of 13. Then in the meantime a user B adds another item to the list. Then user A queries the second page of the list and he'll receive 4 items on the second page, even though the total count will be still cached as 13, because no invalidation was triggered by user A.

Regarding the article you posted, it seems similar to what I had in mind originally, but in addition to that it also invalidates other pages when the total count changes, to prevent flashing the old data (total count), while the new data is being refetched in background. That's also nice to have, although it's basically just a detail in quite rare use case, so not sure if it's worth the effort at the moment, but I'll leave that up to you.

Reply:
Got this. will change the implementation 👍